py/server/deephaven/table_listener.py

```diff
@@ -393,3 +397,205 @@ def listen(t: Table, listener: Union[Callable, TableListener], description: str
+            tables (Sequence[Table]): tables to listen to
+            listener (Union[Callable[[Dict[Table, TableUpdate], bool], None], MergedListener]): listener to process table updates
+                from the tables.
+            description (str, optional): description for the UpdatePerformanceTracker to append to the listener's entry
+            dependencies (Union[Table, Sequence[Table]]): tables that must be satisfied before the listener's execution.
+                A refreshing table is considered to be satisfied if all possible updates to the table have been processed
+                in the current update graph cycle. A static table is always considered to be satisfied. If a specified
+                table is refreshing, it must belong to the same update graph as the table being listened to. Default is
+                None.
+
+                Dependencies ensure that the listener can safely access the dependent tables during its execution. This
+                mainly includes reading the data from the tables. While performing operations on the dependent tables in
+                the listener is safe, it is not recommended because reading or operating on the result tables of those
+                operations may not be safe. It is best to perform the operations on the dependent tables beforehand,
+                and then add the result tables as dependencies to the listener so that they can be safely read in it.
+
+        Raises:
+            DHError
+        """
+        if len(tables) < 2:
+            raise DHError(message="A merged listener must listen to at least 2 refreshing tables.")
+
+        self.tables = tables
+        self.listener_recorders = [_ListenerRecorder(t) for t in tables]
+
+        self.dependencies = dependencies
+
+        if isinstance(listener, MergedListener):
+            self.listener = listener.on_update
+        else:
+            self.listener = listener
+        n_params = len(signature(self.listener).parameters)
+        if n_params != 2:
+            raise ValueError("merged listener function must have 2 parameters (updates, is_replay).")
+
+
+        try:
+            self.merged_listener_adapter = _JPythonMergedListenerAdapter.create(
+                        to_sequence(self.listener_recorders),
+                        to_sequence(self.dependencies),
+                        description,
+                        self)
+            self.started = False
+        except Exception as e:
+            raise DHError(e, "failed to create a merged listener adapter.") from e
+
+    def _process(self, replay_updates: Optional[jpy.JType] = None) -> None:
+        """Process the table updates from the listener recorders.
+
+        Args:
+            replay_updates (Optional[jpy.JType]): the replay updates, only provided during replay.
+        """
+        if replay_updates:
+            self.listener({t: TableUpdate(t, tu) for t, tu in zip(self.tables, j_list_to_list(replay_updates))}, True)
```

Comment:
Just make a separate method for the replay. There's quite literally no value to making `_process` do two things, and it makes the code harder to read. Call `replay()` (or `currentStateAsUpdates()`) to get the list of table updates, then call your callable with the zipped map yourself.

Reply:
Are you suggesting that we make a _replay() for the Java adapter to call? That would require another pyCallable be created.
```
  private PythonMergedListenerAdapter(
            @NotNull ListenerRecorder[] recorders,
            @Nullable NotificationQueue.Dependency[] dependencies,
            @Nullable String listenerDescription,
            @NotNull PyObject pyObjectIn) {
        super(Arrays.asList(recorders), Arrays.asList(dependencies), listenerDescription, null);
        Arrays.stream(recorders).forEach(rec -> rec.setMergedListener(this));
        this.pyCallable = PythonUtils.pyMergeListenerFunc(pyObjectIn); <-------------
    }
```
which isn't much better than reusing _process().

Reply:
No, I'm suggesting that the Python code call a method `List<TableUpdate> currentRowsAsUpdates()`, and call the user's callback directly, rather than making the Java call the Python callback with a more complicated signature, just so the Python callback can branch on the boolean and then call the user's callback.

Reply:
Pseudocode:
```
replay_updates = self.merged_listener_adapter.currentRowsAsUpdates()
try:
    self.listener({t: TableUpdate(t, tu) for t, tu in zip(self.tables, j_list_to_list(replay_updates))}, True)
finally:
    for replay_update in replay_updates:
        replay_update.release()
```